### PR TITLE
Create a mojo interface for LedgerDatabase

### PR DESCRIPTION
**Pull request by zenparsing**
Resolves https://github.com/brave/brave-browser/issues/15779

This PR is an alternative to #8936 that simplifies interaction with `LedgerDatabase` by converting it from a standard C++ abstract class to a mojo interface. Mojo is used to ensure that the implementation instance is only accessed from the dedicated task runner and that callbacks are executed on the originating sequence.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: vendor/brave-ios/Ledger/BATBraveLedger.mm [hunk 1]
```diff
@@ -36,10 +36,13 @@
 #import "base/command_line.h"
 #import "base/i18n/icu_util.h"
 #import "base/ios/ios_util.h"
+
+#include "base/callback_helpers.h"
 #include "base/sequenced_task_runner.h"
 #include "base/task/post_task.h"
 #include "base/task/thread_pool.h"
 #include "base/task_runner_util.h"
+#include "mojo/public/cpp/bindings/remote.h"
```
Comment (kylehickinson) [inline vendor/brave-ios/Ledger/BATBraveLedger.mm:45]: brave-ios BUILD.gn will need to add `//mojo/public/cpp/bindings` as a dep, no?

File: vendor/brave-ios/Ledger/BATBraveLedger.mm [hunk 2]
```diff
@@ -210,9 +196,6 @@ - (void)dealloc
   [NSNotificationCenter.defaultCenter removeObserver:self];
   [self.notificationStartupTimer invalidate];
 
-  if (rewardsDatabase) {
-    databaseQueue->DeleteSoon(FROM_HERE, rewardsDatabase);
-  }
```
Comment (kylehickinson) [inline vendor/brave-ios/Ledger/BATBraveLedger.mm:215]: No manual cleanup required on deconstructor?

**Review by kylehickinson — COMMENTED** [review]
Would like to test this change on iOS before merging, as we've never included a `mojo::Remote` type before and we've had debug crashes in the past due to sequenced task checker

File: vendor/brave-ios/Ledger/BATBraveLedger.mm (hunk 1 above)
Comment (zenparsing) [reply vendor/brave-ios/Ledger/BATBraveLedger.mm:45]: Yep, thanks.

File: vendor/brave-ios/Ledger/BATBraveLedger.mm (hunk 2 above)
Comment (zenparsing) [reply vendor/brave-ios/Ledger/BATBraveLedger.mm:215]: When the `Remote` is destroyed, it's pipe is destroying and the implementation on the other end of the pipe is destroyed (via `mojo::MakeSelfOwnedReceiver`).

**Comment by zenparsing** [comment]
@iefremov Apologies, didn't see your comment about `SequenceBound` until after I posted this PR. Happy to use that abstraction if it's preferred. Generally speaking, I'm looking for techniques that will allow us to do operations across sequence/process boundaries that are dead simple for developers. The `SequenceBound` template does the trick, except all of that `AsyncCall` stuff is unfortunate. We're using mojo here anyway, and I like that it provides a concise way of defining a protocol interface. For sequence safety, my understanding was that passing the pending receiver to the desired sequence, and then setting up the pipe via `mojo::MakeSelfOwnedReceiver` accomplishes what we want. The documentation for mojo isn't stellar, so happy to learn more!

**Comment by iefremov** [comment]
Thanks @zenparsing, correct, I see now that it indeed helps with sequence safety. Though I'd suggest to choose a common approach aka `SequenceBound`:
 * it is a common pattern across the codebase
 * it is easier to understand. It is also explicit about its purpose, while mojo is mostly for IPC or larger component boundaries
 * it requires less crazy boilerplate with binding/remotes/receivers
 * async operations are explicitly marked as `AsyncCall` 
 * it has zero performance overhead

In fact, my plan is to migrate all similar cases to SequenceBound (e.g. the database in ads service, adblock engine, etc), so rewards could be a great first step.

**Comment by zenparsing** [comment]
@iefremov Sounds good - I'll close this PR and update the other one to use `SequenceBound`. Thanks!

**Comment by iefremov** [comment]
thanks @zenparsing !

**Pull request closed without merging**